src/Adapter/CustomerService/QueryHandler/GetCustomerThreadForViewingHandler.php

```diff
@@ -0,0 +1,378 @@
+    }
+
+    /**
+     * @param array $messages
+     *
+     * @return CustomerThreadMessage[]
+     */
+    private function getCustomerThreadMessages(array $messages)
+    {
+        $threadMessages = [];
+
+        foreach ($messages as $key => $message) {
+            $employeeImage = null;
+
+            if ($message['id_employee']) {
+                $employee = new Employee($message['id_employee']);
+                $employeeImage = $employee->getImage();
+            }
+
+            $attachmentFile = null;
+
+            if (isset($message['file_name']) && $message['file_name'] != '') {
+                $attachmentFile = _THEME_PROD_PIC_DIR_ . $message['file_name'];
+            }
+
+            $productId = null;
+            $productName = null;
+
+            if ($message['id_product']) {
+                $product = new Product((int) $message['id_product'], false, $this->context->language->id);
+
+                if (Validate::isLoadedObject($product)) {
+                    $productId = (int) $product->id;
+                    $productName = $product->name;
+                }
+            }
+
+            $type = $message['id_employee'] ? 'employee' : 'customer';
+
+            $threadMessages[] = new CustomerThreadMessage(
+                $type,
+                $message['message'],
+                $message['date_add'],
+                $employeeImage,
+                $message['employee_name'],
+                $message['customer_name'],
+                $attachmentFile,
+                $productId,
+                $productName
+            );
+        }
+
+        return $threadMessages;
+    }
+
+    /**
+     * @param CustomerThreadId $customerThreadId
+     *
+     * @return CustomerThread
+     */
+    private function getCustomerThread(CustomerThreadId $customerThreadId)
+    {
+        $customerThread = new CustomerThread($customerThreadId->getValue());
+
+        if ($customerThread->id !== $customerThreadId->getValue()) {
+            throw new CustomerThreadNotFoundException(
+                sprintf('Customer thread with id "%s" was not found', $customerThreadId->getValue())
+            );
+        }
+
+        return $customerThread;
+    }
+
+    private function getTimeline(array $messages, CustomerThread $customerThread)
+    {
+        $timeline = [];
+
+        foreach ($messages as $message) {
+            $product = new Product((int) $message['id_product'], false, $this->context->language->id);
+
+            $content = '';
+
+            if (!$message['private']) {
+                $content .= sprintf(
+                    '%s <span class="badge badge-primary rounded">%s</span><br/>',
+                    $this->translator->trans('Message to: ', [], 'Admin.Catalog.Feature'),
+                    !$message['id_employee'] ? $message['subject'] : $message['customer_name']
+                );
+            }
+
+            if (Validate::isLoadedObject($product)) {
+                $content .= sprintf(
```

Comment:
@matks I've kept building html in php code, because I want to avoid having complex templates with legacy checks, I think be should improve either now or at some point in the future.

Reply:
Works for now, but as you say we need to clean all of this 👍